manager/shard_strategy.go

```diff
@@ -0,0 +1,191 @@
+		for i := startKey; i < endKey; i++ {
+			container.Args = append(container.Args, "--propeller.include-shard-label", fmt.Sprintf("%d", i))
+		}
+	} else if h.enableUncoveredReplica && podIndex == h.podCount {
+		for i := 0; i < v1alpha1.ShardKeyspaceSize; i++ {
+			container.Args = append(container.Args, "--propeller.exclude-shard-label", fmt.Sprintf("%d", i))
+		}
+	} else {
+		return fmt.Errorf("invalid podIndex '%d' out of range [0,%d)", podIndex, h.GetPodCount())
+	}
+
+	return nil
+}
+
+// Computes a [startKey, endKey) pair denoting the key responsibilities for the provided pod index
+// given the keyspaceSize and podCount parameters
+func computeKeyRange(keyspaceSize, podCount, podIndex int) (int, int) {
+	keysPerPod := keyspaceSize / podCount
+	keyRemainder := keyspaceSize - (podCount * keysPerPod)
+
+	return computeStartKey(keysPerPod, keyRemainder, podIndex), computeStartKey(keysPerPod, keyRemainder, podIndex+1)
+}
+
+func computeStartKey(keysPerPod, keysRemainder, podIndex int) int {
+	return (intMin(podIndex, keysRemainder) * (keysPerPod + 1)) + (intMax(0, podIndex-keysRemainder) * keysPerPod)
+}
+
+func intMin(a, b int) int {
+	if a < b {
+		return a
+	}
+
+	return b
+}
+
+func intMax(a, b int) int {
+	if a > b {
+		return a
+	}
+
+	return b
+}
+
+// The ProjectShardStrategy assigns project(s) to individual FlytePropeller instances to
+// determine FlyteWorkflow processing responsibility.
+type EnvironmentShardStrategy struct {
+	envType                environmentType
+	enableUncoveredReplica bool
+	replicas               [][]string
+}
+
+type environmentType int
+
+const (
+	project environmentType = iota
+	domain
+)
+
+func (e environmentType) String() string {
+	return [...]string{"project", "domain"}[e]
+}
+
+func (e *EnvironmentShardStrategy) GetPodCount() int {
+	if e.enableUncoveredReplica {
+		return len(e.replicas) + 1
+	}
+
+	return len(e.replicas)
+}
+
+func (e *EnvironmentShardStrategy) UpdatePodSpec(pod *v1.PodSpec, podIndex int) error {
+	container, err := getFlytePropellerContainer(pod)
+	if err != nil {
+		return err
+	}
+
+	if podIndex >= 0 && podIndex < len(e.replicas) {
+		for _, entity := range e.replicas[podIndex] {
+			container.Args = append(container.Args, fmt.Sprintf("--propeller.include-%s-label", e.envType), entity)
+		}
+	} else if e.enableUncoveredReplica && podIndex == len(e.replicas) {
+		for _, replica := range e.replicas {
+			for _, entity := range replica {
+				container.Args = append(container.Args, fmt.Sprintf("--propeller.exclude-%s-label", e.envType), entity)
+			}
+		}
+	} else {
+		return fmt.Errorf("invalid podIndex '%d' out of range [0,%d)", podIndex, e.GetPodCount())
+	}
+
+	return nil
+}
+
+func getFlytePropellerContainer(pod *v1.PodSpec) (*v1.Container, error) {
+	// find flytepropeller container(s)
+	var containers []*v1.Container
+	for i := 0; i < len(pod.Containers); i++ {
+		commands := pod.Containers[i].Command
+		if len(commands) > 0 && commands[0] == "flytepropeller" {
```

Comment:
this seems odd, its ok to probably use an annotation or the name itself of the container?

Reply:
updated to use a configurable container name. much cleaner idea.